File: code/python/3495.py
# https://leetcode.com/problems/minimum-operations-to-make-array-elements-zero
MAX_BIT = 16
pref = [0] * MAX_BIT
for i in range(1, MAX_BIT):
    pref[i] = (((3*i - 1) << (2*i)) + 1) // 3

class Solution:
    def minOperations(self, queries: List[List[int]]) -> int:
        res = 0
        for l, r in queries:
            base_l = (l.bit_length() + 1) // 2
            base_r = (r.bit_length() + 1) // 2
            if base_l == base_r:
                cur_res = (r - l + 1) * base_l
            else:
                cur_res = pref[base_r - 1] - pref[base_l] # middle part
                cur_res += (r - ((1 << (2*(base_r - 1))) - 1)) * base_r # right part
                cur_res += (((1 << (2*base_l)) - 1) - l + 1) * base_l # left part
            res += (cur_res + 1) // 2
        return res

File: import/new/3495.java
// https://leetcode.com/problems/minimum-operations-to-make-array-elements-zero
public class Solution {
    public long minOperations(int[][] queries) {
        final int MAX_I = 16;
        long[] pref = new long[MAX_I];
        for (int i = 1; i < MAX_I; i++) {
            pref[i] = ((((long)(3 * i - 1)) << (2 * i)) + 1L) / 3L;
        }

        long res = 0;
        for (int[] q : queries) {
            int l = q[0], r = q[1];

            int base_l = base(l), base_r = base(r);

            long cur_res;
            if (base_l == base_r)
                cur_res = (long)(r - l + 1) * base_l;
            else {
                cur_res = pref[base_r - 1] - pref[base_l];
                cur_res += (long)(r - ((1L << (2 * (base_r - 1))) - 1)) * base_r;
                cur_res += (long)((1L << (2 * base_l)) - l) * base_l;
            }
            res += (cur_res + 1) / 2;
        }
        return res;
    }

    private int base(int x) {
        if (x == 0) return 0;
        int bits = 32 - Integer.numberOfLeadingZeros(x);
        return (bits + 1) / 2;
    }
}
